File: source/errpy/ast.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

type num =
  | Int of int
  | Float of float
  | Complex of float
  | Big_int of string
      (** pyast uses Py.Object.t for Big_int, but this is presumably not what we want (if it's an
          opaque wrapper for a CPython object). Just have the source text representing the integer
          digits for now instead. *)
[@@deriving show]

type object_ = num [@@deriving show]

type constant_desc =
  | Ellipsis
  | Bool of bool
  | Num of num
  | Str of string
  | ByteStr of string
[@@deriving show]

type constant = constant_desc option [@@deriving show]
type singleton = bool option [@@deriving show]

type withitem = {
  context_expr: expr;
  optional_vars: expr option;
}

and recoverableerrorwithlocation = {
  error: string;
  lineno: int;
  col_offset: int;
  end_lineno: int;
  end_col_offset: int;
}

and unaryop =
  | Invert
  | Not
  | UAdd
  | USub

and type_ignore =
  | TypeIgnore of {
      lineno: int;
      tag: string;
    }

and stmt = {
  desc: stmt_desc;
  lineno: int;
  col_offset: int;
  end_lineno: int option;
  end_col_offset: int option;
}

and stmt_desc =
  | FunctionDef of {
      name: string;
      args: arguments;
      body: stmt list;
      decorator_list: expr list;
      type_params: expr list;
      returns: expr option;
      type_comment: string option;
    }
  | AsyncFunctionDef of {
      name: string;
      args: arguments;
      body: stmt list;
      decorator_list: expr list;
      type_params: expr list;
      returns: expr option;
      type_comment: string option;
    }
  | ClassDef of {
      name: string;
      bases: expr list;
      keywords: keyword list;
      body: stmt list;
      decorator_list: expr list;
      type_params: expr list;
    }
  | Return of expr option
  | Delete of expr list
  | Assign of {
      targets: expr list;
      value: expr;
      type_comment: string option;
    }
  | AugAssign of {
      target: expr;
      op: operator;
      value: expr;
    }
  | AnnAssign of {
      target: expr;
      annotation: expr;
      value: expr option;
      simple: int;
    }
  | For of {
      target: expr;
      iter: expr;
      body: stmt list;
      orelse: stmt list;
      type_comment: string option;
    }
  | AsyncFor of {
      target: expr;
      iter: expr;
      body: stmt list;
      orelse: stmt list;
      type_comment: string option;
    }
  | While of {
      test: expr;
      body: stmt list;
      orelse: stmt list;
    }
  | If of {
      test: expr;
      body: stmt list;
      orelse: stmt list;
    }
  | With of {
      items: withitem list;
      body: stmt list;
      type_comment: string option;
    }
  | AsyncWith of {
      items: withitem list;
      body: stmt list;
      type_comment: string option;
    }
  | Match of {
      subject: expr;
      cases: match_case list;
    }
  | Raise of {
      exc: expr option;
      cause: expr option;
    }
  | Try of {
      body: stmt list;
      handlers: excepthandler list;
      orelse: stmt list;
      finalbody: stmt list;
    }
  | Assert of {
      test: expr;
      msg: expr option;
    }
  | Import of alias list
  | ImportFrom of {
      module_: string option;
      names: alias list;
      level: int option;
    }
  | Global of string list
  | Nonlocal of string list
  | Expr of expr
  | Pass
  | Break
  | Continue

and slice = expr

and pattern = {
  desc: pattern_desc;
  lineno: int;
  col_offset: int;
  end_lineno: int;
  end_col_offset: int;
}

and pattern_desc =
  | MatchValue of expr
  | MatchSingleton of constant
  | MatchSequence of pattern list
  | MatchMapping of {
      keys: expr list;
      patterns: pattern list;
      rest: string option;
    }
  | MatchClass of {
      cls: expr;
      patterns: pattern list;
      kwd_attrs: string list;
      kwd_patterns: pattern list;
    }
  | MatchStar of string option
  | MatchAs of {
      pattern: pattern option;
      name: string option;
    }
  | MatchOr of pattern list

and operator =
  | Add
  | Sub
  | Mult
  | MatMult
  | Div
  | Mod
  | Pow
  | LShift
  | RShift
  | BitOr
  | BitXor
  | BitAnd
  | FloorDiv

and mod_ =
  | Module of {
      body: stmt list;
      type_ignores: type_ignore list;
    }
  | Interactive of stmt list
  | Expression of expr
  | FunctionType of {
      argtypes: expr list;
      returns: expr;
    }

and match_case = {
  pattern: pattern;
  guard: expr option;
  body: stmt list;
}

and keyword = {
  arg: string option;
  value: expr;
  lineno: int;
  col_offset: int;
  end_lineno: int option;
  end_col_offset: int option;
}

and expr_context =
  | Load
  | Store
  | Del

and expr = {
  desc: expr_desc;
  lineno: int;
  col_offset: int;
  end_lineno: int option;
  end_col_offset: int option;
}

and expr_desc =
  | BoolOp of {
      op: boolop;
      values: expr list;
    }
  | NamedExpr of {
      target: expr;
      value: expr;
    }
  | BinOp of {
      left: expr;
      op: operator;
      right: expr;
    }
  | UnaryOp of {
      op: unaryop;
      operand: expr;
    }
  | Lambda of {
      args: arguments;
      body: expr;
    }
  | IfExp of {
      test: expr;
      body: expr;
      orelse: expr;
    }
  | Dict of {
      keys: expr option list;
      values: expr list;
    }
  | Set of expr list
  | ListComp of {
      elt: expr;
      generators: comprehension list;
    }
  | SetComp of {
      elt: expr;
      generators: comprehension list;
    }
  | DictComp of {
      key: expr;
      value: expr;
      generators: comprehension list;
    }
  | GeneratorExp of {
      elt: expr;
      generators: comprehension list;
    }
  | Await of expr
  | Yield of expr option
  | YieldFrom of expr
  | Compare of {
      left: expr;
      ops: cmpop list;
      comparators: expr list;
    }
  | Call of {
      func: expr;
      args: expr list;
      keywords: keyword list;
    }
  | FormattedValue of {
      value: expr;
      conversion: int option;
      format_spec: expr option;
    }
  | JoinedStr of expr list
  | Constant of {
      value: constant;
      kind: string option;
    }
  | Attribute of {
      value: expr;
      attr: string;
      ctx: expr_context;
    }
  | Subscript of {
      value: expr;
      slice: expr;
      ctx: expr_context;
    }
  | Starred of {
      value: expr;
      ctx: expr_context;
    }
  | Name of {
      id: string;
      ctx: expr_context;
    }
  | List of {
      elts: expr list;
      ctx: expr_context;
    }
  | Tuple of {
      elts: expr list;
      ctx: expr_context;
    }
  | Slice of {
      lower: expr option;
      upper: expr option;
      step: expr option;
    }

and excepthandler = {
  desc: excepthandler_desc;
  lineno: int;
  col_offset: int;
  end_lineno: int option;
  end_col_offset: int option;
}

and excepthandler_desc =
  | ExceptHandler of {
      type_: expr option;
      name: string option;
      body: stmt list;
    }

and comprehension = {
  target: expr;
  iter: expr;
  ifs: expr list;
  is_async: bool;
}

and cmpop =
  | Eq
  | NotEq
  | Lt
  | LtE
  | Gt
  | GtE
  | Is
  | IsNot
  | In
  | NotIn

and boolop =
  | And
  | Or

and arguments = {
  posonlyargs: arg list;
  args: arg list;
  vararg: arg option;
  kwonlyargs: arg list;
  kw_defaults: expr option list;
  kwarg: arg option;
  defaults: expr list;
}

and arg = {
  arg: string;
  annotation: expr option;
  type_comment: string option;
  lineno: int;
  col_offset: int;
  end_lineno: int option;
  end_col_offset: int option;
}

and alias = {
  name: string;
  asname: string option;
  lineno: int;
  col_offset: int;
  end_lineno: int option;
  end_col_offset: int option;
}
[@@deriving show]

type arguments_args =
  | Expr_list of expr list
  | Arg_list of arg list
[@@deriving show]

type arguments_kwarg =
  | Identifier_opt of string option
  | Arg_opt of arg option
[@@deriving show]

type arguments_vararg =
  | Identifier_opt of string option
  | Arg_opt of arg option
[@@deriving show]

type excepthandler_excepthandler_name =
  | Identifier_opt of string option
  | Expr_opt of expr option
[@@deriving show]

type expr_bytes_s =
  | String of string
  | Bytes of string
[@@deriving show]

type expr_subscript_slice =
  | Slice of slice
  | Expr of expr
[@@deriving show]

type expr_yieldfrom_value =
  | Expr_opt of expr option
  | Expr of expr
[@@deriving show]

type keyword_arg =
  | Identifier_opt of string option
  | Identifier of string
[@@deriving show]

type stmt_importfrom_module =
  | Identifier_opt of string option
  | Identifier of string
[@@deriving show]

File: source/errpy/ast.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#![allow(dead_code)]

use ocamlrep::FromOcamlRep;
use ocamlrep::ToOcamlRep;
use serde::Deserialize;
use serde::Serialize;

#[derive(ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(attr = "deriving show")]
pub struct RecoverableErrorWithLocation {
    pub error: String,
    pub lineno: usize,
    pub col_offset: usize,
    pub end_lineno: usize,
    pub end_col_offset: usize,
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(attr = "deriving show")]
#[repr(C, u8)]
pub enum Num {
    Int(isize),
    Float(f64),
    Complex(f64),
    /// pyast uses Py.Object.t for Big_int, but this is presumably not what we
    /// want (if it's an opaque wrapper for a CPython object). Just have the
    /// source text representing the integer digits for now instead.
    #[rust_to_ocaml(name = "Big_int")]
    BigInt(String),
}

#[rust_to_ocaml(attr = "deriving show")]
pub type Object_ = Num;

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(attr = "deriving show")]
#[repr(C, u8)]
pub enum ConstantDesc {
    Ellipsis,
    Bool(bool),
    Num(Num),
    Str(String),
    ByteStr(String),
}

#[rust_to_ocaml(attr = "deriving show")]
pub type Constant = Option<ConstantDesc>;

#[rust_to_ocaml(attr = "deriving show")]
pub type Singleton = Option<bool>;

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[repr(C)]
pub struct Withitem {
    pub context_expr: Expr,
    pub optional_vars: Option<Expr>,
}

#[derive(Copy, Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(u8)]
pub enum Unaryop {
    Invert,
    Not,
    UAdd,
    USub,
}

#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(C, u8)]
pub enum TypeIgnore {
    TypeIgnore { lineno: isize, tag: String },
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(C)]
pub struct Stmt {
    pub desc: StmtDesc,
    pub lineno: isize,
    pub col_offset: isize,
    pub end_lineno: Option<isize>,
    pub end_col_offset: Option<isize>,
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(C, u8)]
pub enum StmtDesc {
    FunctionDef {
        name: String,
        args: Arguments,
        body: Vec<Stmt>,
        decorator_list: Vec<Expr>,
        type_params: Vec<Expr>,
        returns: Option<Expr>,
        type_comment: Option<String>,
    },
    AsyncFunctionDef {
        name: String,
        args: Arguments,
        body: Vec<Stmt>,
        decorator_list: Vec<Expr>,
        type_params: Vec<Expr>,
        returns: Option<Expr>,
        type_comment: Option<String>,
    },
    ClassDef {
        name: String,
        bases: Vec<Expr>,
        keywords: Vec<Keyword>,
        body: Vec<Stmt>,
        decorator_list: Vec<Expr>,
        type_params: Vec<Expr>,
    },
    Return(Option<Expr>),
    Delete(Vec<Expr>),
    Assign {
        targets: Vec<Expr>,
        value: Expr,
        type_comment: Option<String>,
    },
    AugAssign {
        target: Expr,
        op: Operator,
        value: Expr,
    },
    AnnAssign {
        target: Expr,
        annotation: Expr,
        value: Option<Expr>,
        simple: isize,
    },
    For {
        target: Expr,
        iter: Expr,
        body: Vec<Stmt>,
        orelse: Vec<Stmt>,
        type_comment: Option<String>,
    },
    AsyncFor {
        target: Expr,
        iter: Expr,
        body: Vec<Stmt>,
        orelse: Vec<Stmt>,
        type_comment: Option<String>,
    },
    While {
        test: Expr,
        body: Vec<Stmt>,
        orelse: Vec<Stmt>,
    },
    If {
        test: Expr,
        body: Vec<Stmt>,
        orelse: Vec<Stmt>,
    },
    With {
        items: Vec<Withitem>,
        body: Vec<Stmt>,
        type_comment: Option<String>,
    },
    AsyncWith {
        items: Vec<Withitem>,
        body: Vec<Stmt>,
        type_comment: Option<String>,
    },
    Match {
        subject: Expr,
        cases: Vec<MatchCase>,
    },
    Raise {
        exc: Option<Expr>,
        cause: Option<Expr>,
    },
    Try {
        body: Vec<Stmt>,
        handlers: Vec<Excepthandler>,
        orelse: Vec<Stmt>,
        finalbody: Vec<Stmt>,
    },
    TryStar {
        body: Vec<Stmt>,
        handlers: Vec<Excepthandler>,
        orelse: Vec<Stmt>,
        finalbody: Vec<Stmt>,
    },
    Assert {
        test: Expr,
        msg: Option<Expr>,
    },
    Import(Vec<Alias>),
    ImportFrom {
        module__: Option<String>,
        names: Vec<Alias>,
        level: Option<isize>,
    },
    Global(Vec<String>),
    Nonlocal(Vec<String>),
    Expr(Expr),
    Pass,
    Break,
    Continue,
}

#[rust_to_ocaml(and)]
pub type Slice = Expr;

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(C)]
pub struct Pattern {
    pub desc: Box<PatternDesc>,
    pub lineno: isize,
    pub col_offset: isize,
    pub end_lineno: isize,
    pub end_col_offset: isize,
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(C, u8)]
pub enum PatternDesc {
    MatchValue(Expr),
    MatchSingleton(Constant),
    MatchSequence(Vec<Pattern>),
    MatchMapping {
        keys: Vec<Expr>,
        patterns: Vec<Pattern>,
        rest: Option<String>,
    },
    MatchClass {
        cls: Expr,
        patterns: Vec<Pattern>,
        kwd_attrs: Vec<String>,
        kwd_patterns: Vec<Pattern>,
    },
    MatchStar(Option<String>),
    MatchAs {
        pattern: Option<Pattern>,
        name: Option<String>,
    },
    MatchOr(Vec<Pattern>),
}

#[derive(Copy, Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(u8)]
pub enum Operator {
    Add,
    Sub,
    Mult,
    MatMult,
    Div,
    Mod,
    Pow,
    LShift,
    RShift,
    BitOr,
    BitXor,
    BitAnd,
    FloorDiv,
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(C, u8)]
pub enum Mod_ {
    Module {
        body: Vec<Stmt>,
        type_ignores: Vec<TypeIgnore>,
    },
    Interactive(Vec<Stmt>),
    Expression(Expr),
    FunctionType {
        argtypes: Vec<Expr>,
        returns: Expr,
    },
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(C)]
pub struct MatchCase {
    pub pattern: Pattern,
    pub guard: Option<Expr>,
    pub body: Vec<Stmt>,
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(C)]
pub struct Keyword {
    pub arg: Option<String>,
    pub value: Expr,
    pub lineno: isize,
    pub col_offset: isize,
    pub end_lineno: Option<isize>,
    pub end_col_offset: Option<isize>,
}

#[derive(Copy, Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(u8)]
pub enum ExprContext {
    Load,
    Store,
    Del,
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(C)]
pub struct Expr {
    pub desc: Box<ExprDesc>,
    pub lineno: isize,
    pub col_offset: isize,
    pub end_lineno: Option<isize>,
    pub end_col_offset: Option<isize>,
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(C, u8)]
pub enum ExprDesc {
    BoolOp {
        op: Boolop,
        values: Vec<Expr>,
    },
    NamedExpr {
        target: Expr,
        value: Expr,
    },
    BinOp {
        left: Expr,
        op: Operator,
        right: Expr,
    },
    UnaryOp {
        op: Unaryop,
        operand: Expr,
    },
    Lambda {
        args: Arguments,
        body: Expr,
    },
    IfExp {
        test: Expr,
        body: Expr,
        orelse: Expr,
    },
    Dict {
        // keys is of Optional Expr type because dictionary splats `{**arg1} are
        // represented by a value and a corresponding None key
        // normally of course we expect a non None key and value pair e.g.
        // for `{'a': 12}`
        keys: Vec<Option<Expr>>,
        values: Vec<Expr>,
    },
    Set(Vec<Expr>),
    ListComp {
        elt: Expr,
        generators: Vec<Comprehension>,
    },
    SetComp {
        elt: Expr,
        generators: Vec<Comprehension>,
    },
    DictComp {
        key: Expr,
        value: Expr,
        generators: Vec<Comprehension>,
    },
    GeneratorExp {
        elt: Expr,
        generators: Vec<Comprehension>,
    },
    Await(Expr),
    Yield(Option<Expr>),
    YieldFrom(Expr),
    Compare {
        left: Expr,
        ops: Vec<Cmpop>,
        comparators: Vec<Expr>,
    },
    Call {
        func: Expr,
        args: Vec<Expr>,
        keywords: Vec<Keyword>,
    },
    FormattedValue {
        value: Expr,
        conversion: Option<isize>,
        format_spec: Option<Expr>,
    },
    JoinedStr(Vec<Expr>),
    Constant {
        value: Constant,
        kind: Option<String>,
    },
    Attribute {
        value: Expr,
        attr: String,
        ctx: ExprContext,
    },
    Subscript {
        value: Expr,
        slice: Expr,
        ctx: ExprContext,
    },
    Starred {
        value: Expr,
        ctx: ExprContext,
    },
    Name {
        id: String,
        ctx: ExprContext,
    },
    List {
        elts: Vec<Expr>,
        ctx: ExprContext,
    },
    Tuple {
        elts: Vec<Expr>,
        ctx: ExprContext,
    },
    Slice {
        lower: Option<Expr>,
        upper: Option<Expr>,
        step: Option<Expr>,
    },
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(C)]
pub struct Excepthandler {
    pub desc: ExcepthandlerDesc,
    pub lineno: isize,
    pub col_offset: isize,
    pub end_lineno: Option<isize>,
    pub end_col_offset: Option<isize>,
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(C, u8)]
pub enum ExcepthandlerDesc {
    ExceptHandler {
        type__: Option<Expr>,
        name: Option<String>,
        body: Vec<Stmt>,
        star: bool,
    },
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(C)]
pub struct Comprehension {
    pub target: Expr,
    pub iter: Expr,
    pub ifs: Vec<Expr>,
    pub is_async: bool,
}

#[derive(Copy, Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(u8)]
pub enum Cmpop {
    Eq,
    NotEq,
    Lt,
    LtE,
    Gt,
    GtE,
    Is,
    IsNot,
    In,
    NotIn,
}

#[derive(Copy, Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(u8)]
pub enum Boolop {
    And,
    Or,
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(C)]
pub struct Arguments {
    pub posonlyargs: Vec<Arg>,
    pub args: Vec<Arg>,
    pub vararg: Option<Arg>,
    pub kwonlyargs: Vec<Arg>,
    pub kw_defaults: Vec<Option<Expr>>,
    pub kwarg: Option<Arg>,
    pub defaults: Vec<Expr>,
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and)]
#[repr(C)]
pub struct Arg {
    pub arg: String,
    pub annotation: Option<Expr>,
    pub type_comment: Option<String>,
    pub lineno: isize,
    pub col_offset: isize,
    pub end_lineno: Option<isize>,
    pub end_col_offset: Option<isize>,
}

#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(and, attr = "deriving show")]
#[repr(C)]
pub struct Alias {
    pub name: String,
    pub asname: Option<String>,
    pub lineno: isize,
    pub col_offset: isize,
    pub end_lineno: Option<isize>,
    pub end_col_offset: Option<isize>,
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(attr = "deriving show")]
#[repr(C, u8)]
pub enum ArgumentsArgs {
    #[rust_to_ocaml(name = "Expr_list")]
    ExprList(Vec<Expr>),
    #[rust_to_ocaml(name = "Arg_list")]
    ArgList(Vec<Arg>),
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(attr = "deriving show")]
#[repr(C, u8)]
pub enum ArgumentsKwarg {
    #[rust_to_ocaml(name = "Identifier_opt")]
    IdentifierOpt(Option<String>),
    #[rust_to_ocaml(name = "Arg_opt")]
    ArgOpt(Option<Arg>),
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(attr = "deriving show")]
#[repr(C, u8)]
pub enum ArgumentsVararg {
    #[rust_to_ocaml(name = "Identifier_opt")]
    IdentifierOpt(Option<String>),
    #[rust_to_ocaml(name = "Arg_opt")]
    ArgOpt(Option<Arg>),
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(attr = "deriving show")]
#[repr(C, u8)]
pub enum ExcepthandlerExcepthandlerName {
    #[rust_to_ocaml(name = "Identifier_opt")]
    IdentifierOpt(Option<String>),
    #[rust_to_ocaml(name = "Expr_opt")]
    ExprOpt(Option<Expr>),
}

#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(attr = "deriving show")]
#[repr(C, u8)]
pub enum ExprBytesS {
    String(String),
    Bytes(String),
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(attr = "deriving show")]
#[repr(C, u8)]
pub enum ExprSubscriptSlice {
    Slice(Slice),
    Expr(Expr),
}

#[derive(Clone, Debug, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(attr = "deriving show")]
#[repr(C, u8)]
pub enum ExprYieldfromValue {
    #[rust_to_ocaml(name = "Expr_opt")]
    ExprOpt(Option<Expr>),
    Expr(Expr),
}

#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(attr = "deriving show")]
#[repr(C, u8)]
pub enum KeywordArg {
    #[rust_to_ocaml(name = "Identifier_opt")]
    IdentifierOpt(Option<String>),
    Identifier(String),
}

#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
#[derive(Serialize, Deserialize, ToOcamlRep, FromOcamlRep)]
#[rust_to_ocaml(attr = "deriving show")]
#[repr(C, u8)]
pub enum StmtImportfromModule {
    #[rust_to_ocaml(name = "Identifier_opt")]
    IdentifierOpt(Option<String>),
    Identifier(String),
}
